packages/server-reason-react-ppx/server_reason_react_ppx.ml

```diff
@@ -696,7 +709,183 @@ let props_to_model ~loc (props : (arg_label * expression option * pattern) list)
           [%expr [%e expr] :: [%e acc]])
     props
 
-let rewrite_structure_item structure_item =
+module ServerFunction = struct
+  let rec last_expr_to_fn ~loc expr fn =
+    match expr.pexp_desc with
+    | Pexp_constraint (expr, _) -> last_expr_to_fn ~loc expr fn
+    | Pexp_fun (arg_label, arg_expression, fun_pattern, expression) ->
+        pexp_fun ~loc arg_label arg_expression fun_pattern (last_expr_to_fn ~loc expression fn)
+    | _ -> fn
+
+  let generate_id ~loc name =
+    (* We need to add a nasty hack here, since have different files for native and melange.Assume that the file structure is native/lib and js, and replace the name directly. This is supposed to be temporal, until dune implements https://github.com/ocaml/dune/issues/10630 *)
+    let file_path = loc.loc_start.pos_fname |> Str.replace_first (Str.regexp {|/js/|}) "/native/shared/" in
+    let hash = Printf.sprintf "%s_%s_%d" name file_path loc.loc_start.pos_lnum |> Hashtbl.hash |> string_of_int in
+    hash
+
+  let get_arg_details (arg : arg_label * expression option * pattern) =
```

Comment:
what does this function do? 

Resolve patterns into the core_type, but also validates stuff? I'm confused by this abstraction